app/src/main/java/com/github/mik629/android/fundamentals/data/repositories/MoviesRepositoryImpl.kt

```diff
@@ -1,42 +1,108 @@
 package com.github.mik629.android.fundamentals.data.repositories
 
-import com.github.mik629.android.fundamentals.data.mappers.Mapper
+import com.github.mik629.android.fundamentals.data.db.daos.MovieDao
+import com.github.mik629.android.fundamentals.data.db.models.MovieWithActorsAndGenres
+import com.github.mik629.android.fundamentals.data.db.models.toCrossRef
+import com.github.mik629.android.fundamentals.data.db.models.toEntity
+import com.github.mik629.android.fundamentals.data.db.models.toMovie
 import com.github.mik629.android.fundamentals.data.network.ServerApi
-import com.github.mik629.android.fundamentals.data.network.model.Actor
-import com.github.mik629.android.fundamentals.domain.model.ActorItem
-import com.github.mik629.android.fundamentals.domain.model.GenreItem
-import com.github.mik629.android.fundamentals.domain.model.MovieItem
+import com.github.mik629.android.fundamentals.data.network.model.toActor
+import com.github.mik629.android.fundamentals.data.network.model.toMovie
+import com.github.mik629.android.fundamentals.domain.model.Actor
+import com.github.mik629.android.fundamentals.domain.model.Genre
+import com.github.mik629.android.fundamentals.domain.model.Movie
 import com.github.mik629.android.fundamentals.domain.repositories.MoviesRepository
+import kotlinx.coroutines.CoroutineScope
 import kotlinx.coroutines.Dispatchers
-import kotlinx.coroutines.withContext
+import kotlinx.coroutines.SupervisorJob
+import kotlinx.coroutines.async
+import kotlinx.coroutines.awaitAll
+import kotlinx.coroutines.coroutineScope
+import kotlinx.coroutines.launch
+import timber.log.Timber
+import javax.inject.Inject
 
-class MoviesRepositoryImpl(
+class MoviesRepositoryImpl @Inject constructor(
     private val serverApi: ServerApi,
-    private val actorMapper: Mapper<Actor, ActorItem>
+    private val dao: MovieDao
 ) : MoviesRepository {
+    private val scope: CoroutineScope = CoroutineScope(Dispatchers.IO + SupervisorJob())
 
-    override suspend fun getMovies() = withContext(Dispatchers.IO) {
+    private var imageBaseUrl: String? = null
 
-        val movies = serverApi.getMovieList()
+    override suspend fun getMovies(): List<Movie> {
+        val cachedMovies = getMoviesFromCache()
+        return if (cachedMovies.isEmpty()) {
+            loadMoviesFromNetwork()
+        } else {
+            cachedMovies
+        }
+    }
+
+    override suspend fun getMovie(id: Long): Movie {
+        return getMovieFromCache(id) ?: loadMovieFromNetwork(id = id)
+    }
+
+    private suspend fun getMoviesFromCache(): List<Movie> =
+        dao.getAllMovies()
+            .map(MovieWithActorsAndGenres::toMovie)
+
+    private suspend fun loadMoviesFromNetwork(category: String = "popular"): List<Movie> {
```

Comment:
private suspend fun loadMoviesFromNetwork(category: String = "popular"): List<Movie> =  coroutineScope {
         Timber.d("Loading from the network")
         return serverApi.getMovieList(category)
             .results
             .map { movie -> async { loadMovieFromNetwork(movie.id) } }
             .awaitAll()
             .sortedByDescending { movie -> movie.rating }
             .also { movies -> save(movies) }
     }